packages/cache/src/service.ts

```diff
@@ -0,0 +1,146 @@
+        logger.warn("Corrupted cache data detected, treating as cache miss", {
+          key,
+          parseError,
+        });
+        return err({
+          code: ErrorCode.CacheCorruptionError,
+        });
+      }
+    } catch (error) {
+      logger.error({ error, key }, "Cache get operation failed");
+      return err({
+        code: ErrorCode.RedisOperationError,
+      });
+    }
+  }
+
+  /**
+   * Set a value in cache with automatic JSON serialization and TTL
+   * @param key - Cache key to store under
+   * @param value - Value to store
+   * @param ttlMs - Time to live in milliseconds
+   * @returns Result containing void or an error
+   */
+  async set(key: CacheKey, value: unknown, ttlMs: number): Promise<Result<void, CacheError>> {
+    // Validate both key and TTL in one call
+    const validation = validateInputs([key, ZCacheKey], [ttlMs, ZTtlMs]);
+    if (!validation.ok) {
+      return validation;
+    }
+
+    try {
+      const serialized = JSON.stringify(value);
+      await this.redis.setEx(key, Math.floor(ttlMs / 1000), serialized);
+      return ok(undefined);
+    } catch (error) {
+      logger.error({ error, key, ttlMs }, "Cache set operation failed");
+      return err({
+        code: ErrorCode.RedisOperationError,
+      });
+    }
+  }
+
+  /**
+   * Delete one or more keys from cache (idempotent)
+   * @param keys - Array of keys to delete
+   * @returns Result containing void or an error
+   */
+  async del(keys: CacheKey[]): Promise<Result<void, CacheError>> {
+    // Validate all keys using generic validation
+    for (const key of keys) {
+      const validation = validateInputs([key, ZCacheKey]);
+      if (!validation.ok) {
+        return validation;
+      }
+    }
+
+    try {
+      if (keys.length > 0) {
+        await this.redis.del(keys);
+      }
+      return ok(undefined);
+    } catch (error) {
+      logger.error({ error, keys }, "Cache delete operation failed");
+      return err({
+        code: ErrorCode.RedisOperationError,
+      });
+    }
+  }
+
+  /**
+   * Cache wrapper for functions - implements cache-aside pattern
+   * @param fn - Function to cache
+   * @param key - Cache key
+   * @param ttlMs - Time to live in milliseconds
+   * @returns Result containing cached result or fresh result from function
+   */
+  async withCache<T>(fn: () => Promise<T>, key: CacheKey, ttlMs: number): Promise<Result<T, CacheError>> {
+    // Validate inputs upfront
+    const validation = validateInputs([key, ZCacheKey], [ttlMs, ZTtlMs]);
+    if (!validation.ok) {
+      return validation;
+    }
+
+    try {
+      // Try to get from cache first
+      const cacheResult = await this.get<T>(key);
+      if (!cacheResult.ok) {
+        // Cache operation failed, try to get fresh data
+        logger.debug({ error: cacheResult.error, key }, "Cache get operation failed, fetching fresh data");
+      } else if (cacheResult.data !== null) {
```

Comment:
could there be a case that the data in the cache is deliberately stored as null? since both `JSON.stringify(null)` and `JSON.parse("null")` work fine with null there could be case where we store null in the cache, for example storing null in the cache for a short TTL to avoid some network requests. If this is the case then the cache would be a HIT but the code treat it as a MISS